test/e2e/test/elasticsearch/checks_data.go

```diff
@@ -132,3 +124,33 @@ func (dc *DataIntegrityCheck) Verify() error {
 	}
 	return nil
 }
+
+// dataIntegrityReplicas returns the number of replicas to use for the data integrity check,
+// according to the cluster topology, since it affects the cluster health during the mutation.
+func dataIntegrityReplicas(b Builder) int {
+	initial := b.MutatedFrom
+	if initial == nil {
+		initial = &b // consider mutated == initial
+	}
+
+	if initial.Elasticsearch.Spec.NodeCount() == 1 || b.Elasticsearch.Spec.NodeCount() == 1 {
+		// a 1 node cluster can only be green if shards have no replicas
+		return 0
+	}
+
+	isVersionUpgrade := initial.Elasticsearch.Spec.Version != b.Elasticsearch.Spec.Version
+
+	for _, initialNs := range initial.Elasticsearch.Spec.Nodes {
+		for _, mutatedNs := range b.Elasticsearch.Spec.Nodes {
+			if initialNs.Name == mutatedNs.Name &&
+				(isVersionUpgrade || !reflect.DeepEqual(initialNs, mutatedNs)) {
```

Comment:
deep equalling the nodespec does not catch all cases where we roll the nodes (e.g. secure settings triggered mutations will have identical node specs)

Reply:
Argh, indeed :(
Probably trying to be too smart here, this is becoming complicated.

Maybe we should just pass the number of replicas from the high-level test function directly?

```go
test.RunMutationReversal(t, []test.Builder{b}, []test.Builder{renamed}, test.MutationOptions{Replicas: 1})
```

The "hard" part is for the person writing the test to make sure they pick the right number here, according to the constraints expressed in the PR first post.

Reply:
I like the idea of being explicit, but deciding on the right number of replicas sounds indeed error prone

>Otherwise, use 0 replicas,

Just thinking through the different cases again. What other case would that be that you are referring to? If we are changing config and rolling nodes that is not the 'otherwise' case, so the only thing remaining is single node (I might be missing something ofc) which we already covered. So I am thinking we could simplify the logic to just check the cluster size. 

Also we could fix the problem on the assertion side as well and just run with a default: 
* we could be more flexible and tolerate yellow cluster health during upgrades (we migrate only on downscales anyway)
* during non-HA cluster upgrades temporary 'red' is unavoidable it is just hidden by the fact that we don't get a health check result back, again we could relax the check here

Reply:
After a discussion on Zoom:
* one big other case to cover is addition/deletion/replacement of StatefulSets, where we expect data migration to happen; no rolling upgrade occurs
* we agreed on maybe keeping the PR as is for now, but add a comment explaning how this rolling upgrade detection mechanism does not detect things such as changing secure settings
* the secure settings check does not rely on the generic mutations step anyway